Gateway/Request/ShopperInteractionDataBuilder.php

```diff
@@ -0,0 +1,82 @@
+ * ############    ####( ######  #####. ######  ############   ############
+ * #############  #####( ######  #####. ######  #############  #############
+ *        ######  #####( ######  #####. ######  #####  ######  #####  ######
+ * ###### ######  #####( ######  #####. ######  #####  #####   #####  ######
+ * ###### ######  #####( ######  #####. ######  #####          #####  ######
+ * #############  #############  #############  #############  #####  ######
+ *  ############   ############  #############   ############  #####  ######
+ *                                      ######
+ *                               #############
+ *                               ############
+ *
+ * Adyen Payment module (https://www.adyen.com/)
+ *
+ * Copyright (c) 2021 Adyen NV (https://www.adyen.com/)
+ * See LICENSE.txt for license details.
+ *
+ * Author: Adyen <magento@adyen.com>
+ */
+
+namespace Adyen\Payment\Gateway\Request;
+
+use Adyen\Payment\Model\Ui\AdyenCcConfigProvider;
+use Adyen\Payment\Model\Ui\AdyenOneclickConfigProvider;
+use Adyen\Payment\Model\Ui\AdyenPayByLinkConfigProvider;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\Model\Context;
+use Magento\Payment\Gateway\Helper\SubjectReader;
+use Magento\Payment\Gateway\Request\BuilderInterface;
+
+class ShopperInteractionDataBuilder implements BuilderInterface
+{
+    /**
+     * @var \Magento\Framework\App\State
+     */
+    private $appState;
+
+    /**
+     * RecurringDataBuilder constructor.
+     *
+     * @param Context $context
+     */
+    public function __construct(
+        Context $context
+    ) {
+        $this->appState = $context->getAppState();
+    }
+
+    /**
+     * @param array $buildSubject
+     * @return array
+     * @throws LocalizedException
+     */
+    public function build(array $buildSubject)
+    {
+        /** @var \Magento\Payment\Gateway\Data\PaymentDataObject $paymentDataObject */
+        $paymentDataObject = SubjectReader::readPayment($buildSubject);
+        $payment = $paymentDataObject->getPayment();
+        $paymentMethod = $payment->getMethodInstance()->getCode();
+
+        $requestBody = [];
+        if ($paymentMethod == AdyenCcConfigProvider::CODE &&
+            $this->appState->getAreaCode() == \Magento\Framework\App\Area::AREA_ADMINHTML) {
+            // Backend CC orders are MOTO
+            $requestBody['body']['shopperInteraction'] = "Moto";
+        } elseif ($paymentMethod == AdyenOneclickConfigProvider::CODE
+            || $paymentMethod == AdyenCcConfigProvider::CC_VAULT_CODE) {
+            // OneClick and Vault are ContAuth
+            $requestBody['body']['shopperInteraction'] = 'ContAuth';
+        } elseif ($paymentMethod == AdyenPayByLinkConfigProvider::CODE) {
+            // Don't send shopperInteraction for PBL
+        } else {
+            // Ecommerce is the default shopperInteraction
+            $requestBody['body']['shopperInteraction'] = 'Ecommerce';
+        }
```

Comment:
Hey @cyattilakiss what do you think of this logic? does the shopperInteraction mapping make sense to you?

Reply:
I really like that you've outsourced the shopperInteraction to a separate builder, indeed it was distributed to a lot of places and this is much cleaner now.

The only suggestion I might have if you assign the `shopperInteraction` value to the `$requestBody['body']` at the end of the builder and do an early return when it's PBL it's even a bit easier to manage and understand later. (see in the other suggestion comment)

Reply:
```
if ($paymentMethod == AdyenPayByLinkConfigProvider::CODE) {
    // Don't send shopperInteraction for PBL
    return [];
}

 // Ecommerce is the default shopperInteraction
 $shopperInteraction = 'Ecommerce';
 
if ($paymentMethod == AdyenCcConfigProvider::CODE &&
    $this->appState->getAreaCode() == \Magento\Framework\App\Area::AREA_ADMINHTML) {
    // Backend CC orders are MOTO
    $shopperInteraction = "Moto";
} elseif ($paymentMethod == AdyenOneclickConfigProvider::CODE
    || $paymentMethod == AdyenCcConfigProvider::CC_VAULT_CODE) {
    // OneClick and Vault are ContAuth
    $shopperInteraction = 'ContAuth';
} 
        
$requestBody['body']['shopperInteraction'] = $shopperInteraction;
```

Reply:
Thanks! Added this suggestion